openstack-releases support needed for determining openstack codename

zaza-openstack-tests doesn't currently have wallaby definitions in zaza/openstack/utilities/os_versions.py. What is needed is support similar to that introduced in charm-helpers to use the openstack-release apt package to determine codenames (on releases where the package is available - wallaby+).

```
ubuntu@coreycb-bastion:~/charmed-openstack-tester$ git remote -v
origin  https://github.com/openstack-charmers/charmed-openstack-tester (fetch)
origin  https://github.com/openstack-charmers/charmed-openstack-tester (push)

ubuntu@coreycb-bastion:~/charmed-openstack-tester$ tox -e func-target keystone_v3_smoke_focal:focal-wallaby
...
2021-04-19 15:06:05 [INFO] Checking workload status message of cinder-mysql-router/0                                                                                                                                                                                            
2021-04-19 15:06:05 [INFO] Checking workload status of keystone/0                                                                                                                                                                                                               
2021-04-19 15:06:05 [INFO] Checking workload status message of keystone/0                                                                                                                                                                                                       
2021-04-19 15:06:49 [INFO] looking at application: {'name': 'keystone', 'type': {'pkg': 'keystone', 'origin_setting': 'openstack-origin'}}                                                                                                                                      
2021-04-19 15:06:50 [ERROR] {'keystone_v3_smoke_focal': 'zaza-a42067200f0c'}                                                                                                                                                                                                    
2021-04-19 15:06:50 [ERROR] Model keystone_v3_smoke_focal (zaza-a42067200f0c)                                                                                                                                                                                                   
Traceback (most recent call last):                                                                                                                                                                                                                                              
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/bin/functest-run-suite", line 8, in <module>                             
    sys.exit(main())                                                                                                                                                                                                                                                            
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/charm_lifecycle/func_test_runner.py", line 308, in main                                                                                                                         
    func_test_runner(                                                                                                                                                                                                                                                           
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/charm_lifecycle/func_test_runner.py", line 242, in func_test_runner                                                                                                             
    run_env_deployment(env_deployment, keep_model=preserve_model,                                                                                                                                                                                                               
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/charm_lifecycle/func_test_runner.py", line 141, in run_env_deployment                                                                                                           
    configure.configure(                                                                                                                                                                                                                                                        
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/charm_lifecycle/configure.py", line 51, in configure                                                                                                                            
    run_configure_list(functions)                                                                                                                                                                                                                                               
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/charm_lifecycle/configure.py", line 37, in run_configure_list                                                                                                                   
    utils.get_class(func)()                                                                                                                                                                                                                                                     
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/charm_tests/ceilometer/setup.py", line 30, in basic_setup                                                                                                             
    current_release = openstack_utils.get_os_release()                                                                                                                                                                                                                          
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/utilities/openstack.py", line 1831, in get_os_release                                                                                                                 
    release_pair = get_current_os_release_pair(application=application)                                                                                                                                                                                                         
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/utilities/openstack.py", line 1812, in get_current_os_release_pair                                                                                                    
    os_version = get_current_os_versions([application]).get(application)                                                                                                                                                                                                        
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/utilities/openstack.py", line 1776, in get_current_os_versions                                                                                                        
    get_os_code_info(application['type']['pkg'], version))                                                                              
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/utilities/openstack.py", line 1753, in get_os_code_info                                                                                                               
    return OPENSTACK_CODENAMES[vers]                                                                                                                                                                                                                                            
KeyError: '19.0'


(func-target) ubuntu@coreycb-bastion:~/charmed-openstack-tester/tests/distro-regression$ functest-configure -m keystone_v3_smoke_focal:zaza-a42067200f0c
...
2021-04-19 16:09:20 [INFO] Checking workload status of cinder-mysql-router/0                                                            
2021-04-19 16:09:20 [INFO] Checking workload status message of cinder-mysql-router/0                                                    
2021-04-19 16:09:21 [INFO] Checking workload status of keystone/0                                                                       
2021-04-19 16:09:21 [INFO] Checking workload status message of keystone/0                                                               
2021-04-19 16:10:15 [INFO] looking at application: {'name': 'keystone', 'type': {'pkg': 'keystone', 'origin_setting': 'openstack-origin'}}
Traceback (most recent call last):                                                                                                      
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/bin/functest-configure", line 8, in <module>                             
    sys.exit(main())                                                                                                                    
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/charm_lifecycle/configure.py", line 89, in main
    configure(model_name, funcs)                                                                                                        
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/charm_lifecycle/configure.py", line 51, in configure
    run_configure_list(functions)                                   
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/charm_lifecycle/configure.py", line 37, in run_configure_list
    utils.get_class(func)()                                         
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/charm_tests/ceilometer/setup.py", line 30, in basic_setup
    current_release = openstack_utils.get_os_release()                                                                                  
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/utilities/openstack.py", line 1831, in get_os_release
    release_pair = get_current_os_release_pair(application=application)                                                                 
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/utilities/openstack.py", line 1812, in get_current_os_release_pair
    os_version = get_current_os_versions([application]).get(application)                                                                
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/utilities/openstack.py", line 1776, in get_current_os_versions
    get_os_code_info(application['type']['pkg'], version))                                                                              
  File "/home/ubuntu/charmed-openstack-tester/.tox/func-target/lib/python3.8/site-packages/zaza/openstack/utilities/openstack.py", line 1753, in get_os_code_info
    return OPENSTACK_CODENAMES[vers]                                
KeyError: '19.0'
```

--- a/unit_tests/utilities/test_zaza_utilities_openstack.py
+++ b/unit_tests/utilities/test_zaza_utilities_openstack.py
@@ -884,7 +884,24 @@ class TestOpenStackUtils(ut_utils.BaseTestCase):
         result = openstack_utils.get_current_os_release_pair()
         self.assertEqual(expected, result)
 
-    def test_get_openstack_release(self):
+    def test_get_current_os_versions(self):
+        self.patch_object(openstack_utils, "get_openstack_release")
+        self.patch_object(openstack_utils.generic_utils, "get_pkg_version")
+
+        # Pre-Wallaby scenario where openstack-release package isn't installed
+        self.get_openstack_release.return_value = None
+        self.get_pkg_version.return_value = '18.0.0'
+        expected = {'keystone': 'victoria'}
+        result = openstack_utils.get_current_os_versions('keystone')
+        self.assertEqual(expected, result)
+
+        # Wallaby+ scenario where openstack-release package is installed
+        self.get_openstack_release.return_value = 'wallaby'
+        expected = {'keystone': 'wallaby'}
+        result = openstack_utils.get_current_os_versions('keystone')
+        self.assertEqual(expected, result)
+
+    def test_get_os_release(self):
         self.patch(
             'zaza.openstack.utilities.openstack.get_current_os_release_pair',
             new_callable=mock.MagicMock(),
@@ -935,6 +952,23 @@ class TestOpenStackUtils(ut_utils.BaseTestCase):
         self.get_current_os_versions.return_value = {"keystone": "queens"}
         self.get_application_config_option.return_value = None
         self.assertEqual(openstack_utils.get_keystone_api_version(), 3)
+
+    def test_get_openstack_release(self):
+        self.patch_object(openstack_utils.model, "get_units")
+        self.patch_object(openstack_utils.juju_utils, "remote_run")
+
+        # Test pre-Wallaby behavior where openstack-release pkg isn't installed
+        self.get_units.return_value = []
+        self.remote_run.return_value = "OPENSTACK_CODENAME=wallaby "
+
+        # Test Wallaby+ behavior where openstack-release package is installed
+        unit1 = mock.MagicMock()
+        unit1.entity_id = 1
+        self.get_units.return_value = [unit1]
+        self.remote_run.return_value = "OPENSTACK_CODENAME=wallaby "
+
+        result = openstack_utils.get_openstack_release("application", "model")
+        self.assertEqual(result, "wallaby")
 
     def test_get_project_id(self):
         # No domain

--- a/zaza/openstack/utilities/openstack.py
+++ b/zaza/openstack/utilities/openstack.py
@@ -1770,9 +1770,14 @@ def get_openstack_release(application, model_name=None):
     units = model.get_units(application, model_name=model_name)
     for unit in units:
         cmd = 'cat /etc/openstack-release | grep OPENSTACK_CODENAME'
-        out = juju_utils.remote_run(unit.entity_id, cmd, model_name=model_name)
-        codename = out.split('=')[1].strip()
-        versions.append(codename)
+        try:
+            out = juju_utils.remote_run(unit.entity_id, cmd,
+                                        model_name=model_name)
+        except model.CommandRunFailed:
+            logging.info('Fall back to version check for OpenStack codename')
+        else:
+            codename = out.split('=')[1].strip()
+            versions.append(codename)
     if len(set(versions)) == 0:
         return None
     elif len(set(versions)) > 1:

--- a/zaza/openstack/utilities/openstack.py
+++ b/zaza/openstack/utilities/openstack.py
@@ -1753,8 +1753,43 @@ def get_os_code_info(package, pkg_version):
             return OPENSTACK_CODENAMES[vers]
 
 
+def get_openstack_release(application, model_name=None):
+    """Return the openstack release codename based on /etc/openstack-release.
+
+    This will only return a codename if the openstack-release package is
+    installed on the unit.
+
+    :param application: Application name
+    :type application: string
+    :param model_name: Name of model to query.
+    :type model_name: str
+    :returns: OpenStack release codename for application
+    :rtype: string
+    """
+    versions = []
+    units = model.get_units(application, model_name=model_name)
+    for unit in units:
+        cmd = 'cat /etc/openstack-release | grep OPENSTACK_CODENAME'
+        out = juju_utils.remote_run(unit.entity_id, cmd, model_name=model_name)
+        codename = out.split('=')[1].strip()
+        versions.append(codename)
+    if len(set(versions)) == 0:
+        return None
+    elif len(set(versions)) > 1:
+        raise Exception('Unexpected mix of OpenStack releases for {}: {}',
+                        application, versions)
+    return versions[0]
+
+
 def get_current_os_versions(deployed_applications, model_name=None):
     """Determine OpenStack codename of deployed applications.
+
+    Initially, see if the openstack-release pkg is available and use it
+    instead.
+
+    If it isn't then it falls back to the existing method of checking the
+    version of the package passed and then resolving the version from that
+    using lookup tables.
 
     :param deployed_applications: List of deployed applications
     :type deployed_applications: list
@@ -1769,11 +1804,16 @@ def get_current_os_versions(deployed_applications, model_name=None):
             continue
         logging.info("looking at application: {}".format(application))
 
-        version = generic_utils.get_pkg_version(application['name'],
-                                                application['type']['pkg'],
-                                                model_name=model_name)
-        versions[application['name']] = (
-            get_os_code_info(application['type']['pkg'], version))
+        codename = get_openstack_release(application['name'],
+                                         model_name=model_name)
+        if codename:
+            versions[application['name']] = codename
+        else:
+            version = generic_utils.get_pkg_version(application['name'],
+                                                    application['type']['pkg'],
+                                                    model_name=model_name)
+            versions[application['name']] = (
+                get_os_code_info(application['type']['pkg'], version))
     return versions
 
 

--- a/zaza/openstack/utilities/openstack.py
+++ b/zaza/openstack/utilities/openstack.py
@@ -1774,7 +1774,7 @@ def get_openstack_release(application, model_name=None):
             out = juju_utils.remote_run(unit.entity_id, cmd,
                                         model_name=model_name)
         except model.CommandRunFailed:
-            logging.info('Fall back to version check for OpenStack codename')
+            logging.debug('Fall back to version check for OpenStack codename')
         else:
             codename = out.split('=')[1].strip()
             versions.append(codename)

--- a/zaza/openstack/utilities/os_versions.py
+++ b/zaza/openstack/utilities/os_versions.py
@@ -36,6 +36,7 @@ UBUNTU_OPENSTACK_RELEASE = OrderedDict([
     ('eoan', 'train'),
     ('focal', 'ussuri'),
     ('groovy', 'victoria'),
+    ('hirsute', 'wallaby'),
 ])
 
 
@@ -69,7 +70,8 @@ OPENSTACK_RELEASES_PAIRS = [
     'bionic_queens', 'bionic_rocky', 'cosmic_rocky',
     'bionic_stein', 'disco_stein', 'bionic_train',
     'eoan_train', 'bionic_ussuri', 'focal_ussuri',
-    'focal_victoria', 'groovy_victoria']
+    'focal_victoria', 'groovy_victoria',
+    'focal_wallaby', 'hirsute_wallaby']
 
 SWIFT_CODENAMES = OrderedDict([
     ('diablo',

